pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,229 @@
+            start = time.perf_counter()
+            await callback(message)
+            await ack_queue.put(message.ack_id)
+            if metrics_client:
+                metrics_client.increment('pubsub.consumer.succeeded')
+                metrics_client.histogram('pubsub.consumer.latency.runtime',
+                                         time.perf_counter() - start)
+        except Exception:
+            log.exception('Application callback raised an exception')
+            if metrics_client:
+                metrics_client.increment('pubsub.consumer.failed')
+        finally:
+            semaphore.release()
+
+    async def consumer(
+            message_queue: MessageQueue,
+            callback: ApplicationHandler,
+            ack_queue: 'asyncio.Queue[str]',
+            ack_deadline_cache: AckDeadlineCache,
+            consumer_pool_size: int,
+            metrics_client: Optional[MetricsAgent]) -> None:
+        semaphore = asyncio.Semaphore(consumer_pool_size)
+        while True:
+            await semaphore.acquire()
+
+            message, pulled_at = await message_queue.get()
+
+            ack_deadline = await ack_deadline_cache.get()
+            if (time.perf_counter() - pulled_at) >= ack_deadline:
+                if metrics_client:
+                    metrics_client.increment('pubsub.consumer.failfast')
+                message_queue.task_done()
+                semaphore.release()
+                continue
+
+            if metrics_client:
+                metrics_client.histogram(
+                    'pubsub.consumer.latency.receive',
+                    time.time() - message.publish_time.timestamp())
+
+            asyncio.ensure_future(_fire_callback(
+                semaphore,
+                message,
+                callback,
+                ack_queue,
+                metrics_client,
+            ))
+            message_queue.task_done()
+
+    async def producer(
+            subscription: str,
+            message_queue: MessageQueue,
+            subscriber_client: 'SubscriberClient',
+            max_messages: int,
+            metrics_client: Optional[MetricsAgent]
+    ) -> None:
+        while True:
+            try:
+                new_messages = await subscriber_client.pull(
+                    subscription=subscription, max_messages=max_messages)
+            except asyncio.TimeoutError:
+                continue
+
+            pulled_at = time.perf_counter()
+            for m in new_messages:
+                await message_queue.put((m, pulled_at))
+
+            if metrics_client:
+                metrics_client.histogram(
+                    'pubsub.producer.batch', len(new_messages))
+
+            await message_queue.join()
+
+    async def subscribe(subscription: str,  # pylint: disable=too-many-locals
+                        handler: ApplicationHandler,
+                        *,
+                        num_workers: int = 1,
+                        max_messages: int = 100,
```

Comment:
@shaundialpad @TheKevJames we were discussing whether `max_messages` should stay as is to signal its origin in API's `maxMessages` parameter.